source/common/network/happy_eyeballs_connection_impl.h

```diff
@@ -0,0 +1,160 @@
+  Ssl::ConnectionInfoConstSharedPtr ssl() const override;
+  State state() const override;
+  bool connecting() const override;
+  void write(Buffer::Instance& data, bool end_stream) override;
+  void setBufferLimits(uint32_t limit) override;
+  uint32_t bufferLimit() const override;
+  bool aboveHighWatermark() const override;
+  const ConnectionSocket::OptionsSharedPtr& socketOptions() const override;
+  absl::string_view requestedServerName() const override;
+  StreamInfo::StreamInfo& streamInfo() override;
+  const StreamInfo::StreamInfo& streamInfo() const override;
+  absl::string_view transportFailureReason() const override;
+  bool startSecureTransport() override;
+  absl::optional<std::chrono::milliseconds> lastRoundTripTime() const override;
+  void addConnectionCallbacks(ConnectionCallbacks& cb) override;
+  void removeConnectionCallbacks(ConnectionCallbacks& cb) override;
+  void close(ConnectionCloseType type) override;
+  Event::Dispatcher& dispatcher() override;
+  uint64_t id() const override;
+  void hashKey(std::vector<uint8_t>& hash) const override;
+  void setConnectionStats(const ConnectionStats& stats) override;
+  void setDelayedCloseTimeout(std::chrono::milliseconds timeout) override;
+
+  // ScopeTrackedObject
+  void dumpState(std::ostream& os, int indent_level) const override;
+
+private:
+  // ConnectionCallbacks which will be set on an ClientConnection which
+  // sends connection events back to the HappyEyeballsConnectionImpl.
+  class ConnectionCallbacksWrapper : public ConnectionCallbacks {
+  public:
+    ConnectionCallbacksWrapper(HappyEyeballsConnectionImpl& parent, ClientConnection& connection)
+        : parent_(parent), connection_(connection) {}
+
+    void onEvent(ConnectionEvent event) override { parent_.onEvent(event, this); }
+
+    void onAboveWriteBufferHighWatermark() override {
+      parent_.onAboveWriteBufferHighWatermark(this);
+    }
+
+    void onBelowWriteBufferLowWatermark() override { parent_.onBelowWriteBufferLowWatermark(this); }
+
+    // Not needed? interesting.
+    ClientConnection& connection() { return connection_; }
+
+  private:
+    HappyEyeballsConnectionImpl& parent_;
+    ClientConnection& connection_;
+  };
+
+  std::unique_ptr<ClientConnection> createNextConnection();
+  void tryAnotherConnection();
+  void maybeScheduleNextAttempt();
+
+  void onEvent(ConnectionEvent event, ConnectionCallbacksWrapper* wrapper);
+
+  void onAboveWriteBufferHighWatermark(ConnectionCallbacksWrapper* wrapper);
+
+  void onBelowWriteBufferLowWatermark(ConnectionCallbacksWrapper* wrapper);
+
+  void onWriteBufferHighWatermark();
+
+  void cleanupWrapperAndConnection(ConnectionCallbacksWrapper* wrapper);
+
+  // State which needs to be applie to every connection attempt.
+  struct PerConnectionState {
+    absl::optional<bool> detect_early_close_when_read_disabled_;
+    absl::optional<bool> no_delay_;
+    absl::optional<const ConnectionStats*> connection_stats_;
```

Comment:
Could you elaborate on why `absl::optional<const ConnectionStats*>` is used here? I would imagine using `OptRef<const ConnectionStats>` to be compatible with the Envoy style.

Reply:
Hm....... Ooo! Ok, so this is a great question! So there are potentially multiple callbacks so we need a vector<Callback> instead of a single Callback. The optional is that we may have 0 callbacks in case we don't want to have the vector populated at all. I started with vector<Callback&> but it turns out that C++ sucks! In order to have a vector<T> you need to have a default constructable T. But C++ refs are not default constructable. *sigh*

WDYT?

Reply:
Hmm I think we are looking at different places here. My comment is on `PerConnectionState::connection_stats_` but not `PostConnectState::connection_callbacks_`

Reply:
Oh good golly! You're totally right. *facepalm*

Done!